src/builder.js

```diff
@@ -1,43 +1,61 @@
-var _ = require('lodash');
-var xmlBuilder = require('xmlbuilder');
+// @ts-check
 var path = require('path');
 var makeDir = require('make-dir');
 var fs = require('fs');
-var TestSuite = require('./test_suite');
-var TestCase = require('./test_case');
+var { TestSuites } = require('./test_suites');
 
-function JUnitReportBuilder(factory) {
-  this._factory = factory;
-  this._testSuitesAndCases = [];
+class JUnitReportBuilder {
+  /**
+   * @param {import('./factory').Factory} factory
+   */
+  constructor(factory) {
+    this._factory = factory;
+    this._rootTestSuites = new TestSuites(factory);
+  }
+
+  /**
+   * @param {string} reportPath
+   */
+  writeTo(reportPath) {
+    makeDir.sync(path.dirname(reportPath));
+    fs.writeFileSync(reportPath, this.build(), 'utf8');
+  }
+
+  /**
+   * @returns {string}
+   */
+  build() {
+    var xmlTree = this._rootTestSuites.build();
+    return xmlTree.end({ pretty: true });
+  }
+
+  /**
+   * @returns {import('./test_suites').TestSuites}
+   */
+  testSuites() {
+    return this._rootTestSuites;
+  }
```

Comment:
I see a slight risk of confusion here, since this is a getter and not part of the factory/builder pattern like the others, and also since "test suites" may refer both to all test suites and the `<testsuites>` element..

The spec I've previosly relied on does not include any attributes for `<testsuites>` but [this one does](https://github.com/testmoapp/junitxml). The only one that makes sense to expose for custom non-calculated values is `name`. I'm leaning towards introducing a `name(value)` method here instead, that in turn calls `this._rootTestSuites.name(value)`.

(I realize that this would make the `JUnitReportBuilder` class a complete wrapper of the `TestSuites` class. Perhaps they could be the merged at some point.)

I'll fix this.

Reply:
I'm not 100% certain it got better, but I did this change.